### ISSUE TITLE
torch.multinomial selects elements with zero weight

### ISSUE THREAD

**Issue by bderrett** [pytorch/pytorch#48841]
## 🐛 Bug

torch.multinomial sometimes selects elements with zero weight.

## To Reproduce

When I run this code, element 0 is often selected within 200 iterations on CPU and within 1000 iterations on CUDA:

```
for i in range(10000):
    print(i)
    probabilities = torch.zeros(100000, 2)
    probabilities[:, 1] = 1.0
    result = torch.multinomial(probabilities, num_samples=1)[:, 0]
    assert (result == 1).all()
```

## Expected behavior

An element with weight 0 is never selected.

## Environment

PyTorch version: 1.7.0+cu101
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Quadro GV100
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.0+cu101
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.2.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.1           py38hbc911f0_0
[conda] numpy-base                1.19.1           py38hfa32c7d_0
[conda] torch                     1.8.0a0                  pypi_0    pypi

## Additional context
See also https://github.com/pytorch/pytorch/issues/4858 and https://github.com/pytorch/pytorch/issues/13867

Many thanks for all your work on PyTorch! I find it very useful.

cc @ezyang @gchanan @zou3519 @seemethere @malfet @fritzo @neerajprad @alicanb @nikitaved @pbelevich @bdhirsh @jbschlosser @walterddr

**Comment by malfet** [comment]
Cannot reproduce it on machine running FC30, CUDA-10.1 (not that it matters in this case), Python-3.7  and PyTorch-1.7.0
@bderrett can you please share a bit more information about your system?

**Comment by malfet** [comment]
Modified above example in a way that it reproduces in both 1.7 and 1.8:
```
#!/usr/bin/env python
import torch
import time
for i in range(10000):
    probabilities = torch.zeros(100000, 3)
    probabilities[:, 1] = 0.5
    probabilities[:, 2] = 0.7
    start = time.time()
    result = torch.multinomial(probabilities, num_samples=2)
    end = time.time()
    if result.min() == 0:
        assert False, f"{i}: {result} is a failure.\n    result[{result.argmin()>>1}]={result[result.argmin()>>1]}"
    print(f"{i}: {end-start}")
```

**Comment by malfet** [comment]
Will follow up with @ngimel on how reliable is the failure. (I can reproduce it on my Linux box, but never on a Mac, so can it be a compiler problem?)

**Comment by walterddr** [comment]
same here. unable to repro on mac

**Comment by matejklemen** [comment]
Hi, I'm encountering the same problem, so I'm leaving my example here in case it helps, although I see now that it's quite similar to what you've posted above.

`sample_permutations` randomly samples permutations of `indices`. I ran the following script on Colab and it printed out some erroneous samples quite fast.

```python
import torch
def sample_permutations(upper: int, indices: torch.Tensor, num_permutations: int):
    probas = torch.zeros((num_permutations, upper))
    probas[:, indices] = 1 / indices.shape[0]
    return torch.multinomial(probas, num_samples=indices.shape[0])

perturbable_inds = torch.tensor([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29])
num_features = 41
num_samples = 1

for _ in range(10_000_000):
    sampled = sample_permutations(num_features, perturbable_inds, num_samples)
    if torch.any(torch.logical_or(sampled < 1, sampled > 29)):
        print(sampled)
```

Output (36 and 37 shouldn't be sampled):
```
tensor([[18,  4, 20, 25, 13,  6,  3, 28, 27,  2, 10, 22,  5, 15, 26,  7, 12, 17,
         29, 16,  9, 14, 21, 11,  8,  1, 24, 36]])
tensor([[ 6,  2, 13, 18, 16, 11, 15, 22,  7, 26,  8, 10, 17,  9,  5, 29, 25, 27,
         20, 24,  4,  3, 12, 23, 14,  1, 28, 37]])
...
```

The colab env:
```
PyTorch version: 1.7.0+cu101
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.7.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.8.1+cu101
[conda] Could not collect
```

**Comment by binshengliu** [comment]
I ran into the same problem. I saved the prob tensor and I can almost 100% reproduce it using the saved tensor on GPUs and on CPUs.

```
(Pdb++) prob
tensor([0.0000e+00, 2.1031e-35, 0.0000e+00,  ..., 0.0000e+00, 0.0000e+00,
        0.0000e+00], device='cuda:2', grad_fn=<SelectBackward>)
(Pdb++) prob.sum()
tensor(1., device='cuda:2', grad_fn=<SumBackward0>)
(Pdb++) prob.nonzero().shape
torch.Size([53, 1])
(Pdb++) torch.multinomial(prob, 4)
tensor([ 7694, 57959,     1,     0], device='cuda:2')
(Pdb++) prob[[ 7694, 57959,     1,     0]]
tensor([9.0748e-01, 9.2524e-02, 2.1031e-35, 0.0000e+00], device='cuda:2',
       grad_fn=<IndexBackward>)
(Pdb++) prob[0] == 0
tensor(True, device='cuda:2')
(Pdb++) prob_cpu = prob.cpu()
(Pdb++) torch.multinomial(prob_cpu, 4)
tensor([ 7694, 57959,     1,     3])
(Pdb++) prob_cpu[3]==0
tensor(True)
```

My environment:
```
PyTorch version: 1.7.0+cu110
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: 8.0.0 (tags/RELEASE_800/final)
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: 
GPU 0: Tesla V100-DGXS-32GB
GPU 1: Tesla V100-DGXS-32GB
GPU 2: Tesla V100-DGXS-32GB
GPU 3: Tesla V100-DGXS-32GB

Nvidia driver version: 450.80.02
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.4
[pip3] numpy-indexed==0.3.5
[pip3] pytorch-lightning==1.1.0-backports
[pip3] torch==1.7.0+cu110
[conda] numpy                     1.18.4                   pypi_0    pypi
[conda] numpy-indexed             0.3.5                    pypi_0    pypi
[conda] pytorch-lightning         1.1.0-backports           dev_0    <develop>
[conda] torch                     1.7.0+cu110              pypi_0    pypi
```

**Comment by ezyang** [comment]
Re-increasing priority based on user activity

**Comment by ngimel** [comment]
@binshengliu please try on master, there were fixes after 1.7.

**Comment by binshengliu** [comment]
Thanks for the reply. I can confirm my issue was gone on the nightly build `1.8.0.dev20210107+cu110`. May I ask what the problem was? Which stable version will contain the fixes?

**Comment by ngimel** [comment]
It was fixed in #47387, there was a relatively rare problem where "0" was generated at the same position as a non-zero probability, effectively masking it.

**Comment by binshengliu** [comment]
I recently came into this issue again using the latest nightly build. It only occurred when the probs are float16. All good with float32.

```python
import torch

print(torch.__version__)
probs = torch.zeros((10000,), device="cuda", dtype=torch.float16)
probs[0] = 1.0
for i in range(1000):
    value = torch.multinomial(probs, 1)
    if value != 0:
        print(i, value)
        break
```

```bash
$ python test.py
1.8.0.dev20210208+cu110
147 tensor([7358], device='cuda:0')
```

**Comment by ngimel** [comment]
Reopening, can repro on master

**Comment by raquelhortab** [comment]
Hi, I think I am dealing with this too, but from the `WeightedRandomSampler`, which calls `multinomial`.

**Context**: I was looking for a way to combine the functionalities of WeightedRandomSampler and SubsetRandomSampler, since I got 1 dataset and I was using SubsetRandomSampler to handle train and validation splits. However, I also need to balance the train split but, sadly, it seems there's no sampler that can do both things... Therefore, one idea was to use WeightedRandomSampler for the train split and set weight = 0 for the validation indices. This felt a bit risky so I performed a small test to confirm it would never generate samples that have weight=0. This is what I did:

```
sampler = WeightedRandomSampler(weights=[0]*(len(ds)-1) + [1], num_samples=10, replacement=False)
data_loader = DataLoader(ds, batch_size=16, num_workers=1, sampler=sampler)
expected_data = ds[-1]
ok_count = 0
for data in data_loader:
  if torch.equal(data["ids"], expected_data["ids"].repeat(len(data["ids"]),1)):
      print("ok")
      ok_count += 1
  else:
      print(data["ids"])
      print(expected_data["ids"].repeat(len(data["ids"]),1))
      raise Exception("Unexpected data")
print("ok: ", ok_count)
```
and got this:
```
tensor([[ 1, 15, 16,  ...,  0,  0,  0],
        [ 1, 15,  9,  ...,  0,  0,  0],
        [ 1, 15, 16,  ...,  0,  0,  0],
        ...,
        [ 1, 15,  9,  ...,  0,  0,  0],
        [ 1, 15, 15,  ...,  0,  0,  0],
        [ 1, 15, 15,  ...,  0,  0,  0]])
tensor([[ 1, 15, 16,  ...,  0,  0,  0],
        [ 1, 15, 16,  ...,  0,  0,  0],
        [ 1, 15, 16,  ...,  0,  0,  0],
        ...,
        [ 1, 15, 16,  ...,  0,  0,  0],
        [ 1, 15, 16,  ...,  0,  0,  0],
        [ 1, 15, 16,  ...,  0,  0,  0]])
```
It did use different samples but only 1 had weight = 1.
This did not happen when I used `replacement=True` but it seems when you don’t let it replace, it needs to get other samples, maybe?  

`torch. __version__` returns `'1.13.1+cu117'`

**Comment by raquelhortab** [comment]
Sorry to continue writing here but this really worries me as using WeightedRandomSampler could be leading to data leaks in model training and makes me wanna look for an entirely different solution....However, these samplers are quite convenient if they work as expected. Could someone from the development team/community confirm for sure that there is no way a sample with weight 0 could be used if we set `replacement=True` ? Thanks!

**Comment by ngimel** [comment]
Yes, with `replacement=True` a different algorithm is used, and samples with weight = 0 are never generated. However, `replacement=False` doesn't check if the number of non-zero weights is enough to satisfy the number of requested samples, and can get into zero territory.

**Comment by raquelhortab** [comment]
@ngimel thanks for answering! Still, the method does raise an exception when you pass all 0 weights since it cannot generate any sample. I believe it should also raise an exception if it cannot generate k samples given the specified weights. Otherwise it feels like an inconsistent behaviour.

**Comment by myazdani** [comment]
I am also seeing this issue come up in google colab often, for example: https://colab.research.google.com/drive/1HfoMK-jKTYBNpsBjaEncfw1eDUs2Ftm2?usp=sharing

**Comment by myazdani** [comment]
> 

I have further noticed this problem only happens on CPU (and with float data type), not on GPU (regardless of datatype). 

Any ideas @ngimel ?

**Comment by ngimel** [comment]
There's a comment in the code that cpu results are problematic https://github.com/pytorch/pytorch/blob/925a3788ec5c06db62ca732a0e9425a26a00916f/aten/src/ATen/native/Distributions.cpp#L631-L637, and I suspect somehow the probability of getting 0 from exponential_ in cpu is `2**-24`, not `2**-53`

**Comment by HarveyYan** [comment]
the problem is back on torch 2.0.0. Categorical samples positions whose logits have been masked to -inf. Only happens on GPUs.

**Comment by sgugger** [comment]
The issue does seem to have come back on PyTorch 2.0: the following
```py
import torch

torch.manual_seed(0)
probs = torch.tensor([1.0] + [0.0] * 999)
wrongs = 0
for i in range(1000000):
    sampled = torch.multinomial(probs, num_samples=1)
    if sampled != 0:
        wrongs += 1

print(f"{100 * wrongs / 1000000:.4f}")
```
prints a perfect 0 on 1.13.0 but 0.0042 on 2.0

**Comment by ngimel** [comment]
Looks someone broke `exponential_` probably trying to speed it up, `(torch.empty(100000000).exponential_() == 0).sum() ` returns non-zero on cpu, whereas if it indeed was 2**-53 probability this wouldn't be happening. @sgugger can you try the exponential line above on 1.13 if you have it ready?
cc @lezcano @min-jean-cho

**Comment by min-jean-cho** [comment]
If MKL is enabled, the MKL cpu exponential kernel is defined for `x >= 0`, so 0 is not excluded: https://github.com/pytorch/pytorch/blob/e88e92e7a20afb406dc422180a7cf5487fde254c/aten/src/ATen/native/cpu/DistributionKernels.cpp#L151-L152

**Comment by ngimel** [comment]
Then you probably should filter `exponential_` that's computated in multinomial generation

**Comment by min-jean-cho** [comment]
May I know if the choice of exponential for multinomial was based on performance?
Regardless of multinomial, I'll provide a fix for MKL `exponential_` to exclude 0. cc. @jgong5

**Comment by ngimel** [comment]
The linked comment explains algorithm used for generation of multinomial without replacement, it involves computing exponential distribution.

**Comment by sgugger** [comment]
@ngimel, `(torch.empty(100000000).exponential_() == 0).sum()` does return `tensor([0.])` on PyTorch 1.13 as you suspected.

**Comment by hmk114** [comment]
Same bug on python version 3.10.11+pytorch version 2.0.0+windows+cpu
The following code will raise assertion error:
` import torch
p=torch.tensor([1.]+[0.]*1023)
while True:
        assert torch.multinomial(p,1)==0`

**Comment by min-jean-cho** [comment]
I verified the root-cause of this is because MKL exponential kernel is defined for `x >= 0`. Instead of using exponential distribution, why not use uniform distribution as from the original linked comment:

```
X ~ Multinomial(probs, k)
X = topk( log(U) / probs ), where U ~ Uniform(0, 1) 
```

Since `torch.tensor.uniform_` is defined for `[0, 1)`, we can use `1 - U` instead for `(0, 1]` to exclude 0.

**Comment by lezcano** [comment]
why not sample the exponential again in the unlikely case we sample 0?

**Comment by min-jean-cho** [comment]
I had a thought, and I think we can just simply add a constant value to the denominator to ensure that the denominator should not be zero (when CPU MKL exponential kernel samples exactly 0.0) :
```
X = argmax(P/E), where E ~ Exponential(1)
  = argmax(P/(E+c)), where c > 0
```

```python
import torch

def multinomial(probs, num_samples):
    c = 1.0 # c > 0, just use c = 1.0 for convenience
    E = torch.empty_like(probs).exponential_(1)
    return ( probs / (E+c) ).topk(k=num_samples).indices

torch.manual_seed(0)
probs = torch.tensor([1.0] + [0.0] * 999)
wrongs = 0
for i in range(1000000):
    #sampled = torch.multinomial(probs, num_samples=1)
    sampled = multinomial(probs, 1)
    if sampled != 0:
        wrongs += 1

print(wrongs) # 0
```

**Comment by ngimel** [comment]
This is wrong - 
1) in the general case we get not just argmax, but topk elements instead, and this equality wouldn't hold
2) even if it's just argmax, P is an array, and the order of p1/(e1+c) and p2/(e2+c) could be different from the order of p1/e1 and p2/e2

**Comment by min-jean-cho** [comment]
Adding `c` to `e` would not affect argmax operation -- it does not change the direction of increasement or decreasement. 
```
argmax(1/z) = argmax(1/(z+1))
                         ^ it is just scaling
```

**Comment by ngimel** [comment]
This is a duplicate of #50034, we should still decide if we want safe/unsafe versions of multinomial that would ensure that there are enough non-zero weights for sampling without replacement

**Comment by min-jean-cho** [comment]
Please see my comment in #50034. `torch.multinomial` has few confusing problems. Currently, the documentation says `torch.multinomial` samples from multinomial distribution, but it does not (I think `torch.multinomial` is converting sample point from multinomial distribution to categorical somewhere).  We need to align on whether `torch.multinomial` is like `np.random.choice` or `np.random.multinomial`.

**Comment by min-jean-cho** [comment]
Also, if `torch.multinomial` is like `np.random.multinomial`, the probs must sum to 1. That is the definition of multinomial distribution.
